### ISSUE TITLE
Loading plugins when executing a command

### ISSUE THREAD

**Issue by jarra-mcintyre** [bitwalker/distillery#182]
I'm using Conform+Distillery (great work on them btw!) for an umbrella app which accesses a Postgresql DB. The DB connection string .etc. is stored in my .conf file.

I initially followed the Distillery guide for setting up a migration command for running in production. However, this does not work, as it seems that Distillery does not load plugins when executing commands (i.e. running ./bin/myapp command ....) and in fact even with an Erlang config file generated by conform in the var folder the configuration is not loaded.

I was able to run the migration by starting my app, connecting a remote console and executing Ecto.Migrator.run directly. It wouldn't be hard to automate this via an RPC. All that would be required is to strip away the logic for loading the app from the migration command, but this prevents running the migration before starting the app. If the application depends on the DB to be in a particular state to run, this may in the worst case scenario prevent the application from loading.

I have 4 questions / issues:

1. Is there a better way to achieve this than sending an RPC call to a running app?

2. Is there a way to manual load a conform configuration from a command? (Perhaps this should be a seperate question/issue on the Conform project)

3. If there is a way to achieve this use case, perhaps the documentation wrt. to setting up commands, in particular the migration command example should be updated.

3. If not, perhaps the documentation should be updated to reflect this limitation.

I'd be happy to update the documentation, if I knew the best strategy for handling this use case.

**Comment by bitwalker** [comment]
The problem you encountered is due to a couple things which probably need to be documented better:

- Using the `command` task runs the VM with a minimal environment - only `kernel`, `stdlib`, `compiler`, `elixir`, and `iex` are loaded/started. The code path is populated with paths to all the other applications in the release, but they are not loaded/started, you must explicitly do so with `Application.load/1` and `Application.ensure_all_started/1` (if you need it started). For example, if you have code in an app called `mylib`, you would need to execute `:ok = Application.load(:mylib)` in order to access modules from `mylib`.
- Configuration via `Application.get_env/get_all_env` is only available if the application is started. This is enforced by the VM, so there isn't anything we can do about this, but there is a workaround. Namely loading the config file from disk, with `{:ok, config} = :file.consult(~c(#{path_to_sys_config}))`. `config` in this case will be bound to the datastructure defined in `sys.config`, e.g. something like `[myapp: [foo: :bar, ...], ...]`. You can then use standard library functions to extract what you need from it.

The above means it's a bit of a pain to setup standalone tasks which also use code from your application. There are a couple things you can do to make it easier, like putting code for such tasks in their own application which can be trivially loaded/started prior to use. This is also why the custom commands feature of Distillery is provided, as you can setup the VM to automatically load/start the things you want similar to how the `command` task works, and reuse it for multiple things. The documentation covers how to set up a custom command for running migrations, but perhaps I need to review it to ensure I didn't miss the stuff we're talking about now.

To answer your questions:

1. Yes, basically using the techniques I described above you can write the custom code you need for your task once, and reuse it for multiple tasks, i.e. something like `bin/myapp tools migrate`, where tools is the custom command containing the shared setup code, and migrate tells the custom command to call a specific function or module.
2. You can run the conform binary on it's own as it's bundled in the release, and have it spit out the transformed sys.config to a place where you can read it as I described above, but this isn't ideal. I'll try and think of a way I can make this easier. It's a mixed conform/distillery question I think, so it's fine to keep it open here for now.
3. Answered this above, but yes.
4. I think the documentation probably is missing some things regardless, so I'll make sure that is addressed.

Hopefully the above helps you in the short term, let me know if there is anything I can describe better, or if you are still running into problems.

**Comment by jarra-mcintyre** [comment]
Hi, thanks for the quick response. When I get the chance I'm going to try porting over the conform pre-start script to generate a sys.config file and then pass the path to my migration task so I can load it.

Once I get that done, I'd be happy to update the documentation and send a pull request :)

**Comment by spscream** [comment]
@bitwalker I think there is related issue with conform, it overwrites custom sys.config file:
https://github.com/bitwalker/conform/issues/108

**Comment by GalaxyGorilla** [comment]
@jarra-mcintyre I just stumbled upon this issue. It would be really nice to have conform and distillery working together with `command`, especially for migrations. Did you figure out some good way to do this using the conform release plugin somehow?

**Comment by bitwalker** [comment]
@GalaxyGorilla Just FYI, I'll be working on sorting out the conform/distillery integration issues starting this week. I'll update the relevant issues in both repos as they are handled.

**Comment by spscream** [comment]
@bitwalker fixed for me with 2.1.2 conform and current distillery master. Thanks!

**Comment by bitwalker** [comment]
Great :), thanks for the quick response!

**Comment by umatomba** [comment]
hi @bitwalker. 

Do you have some ideas/changes regarding the conform/distillery question?
If i use command for migration triggering, i still have problems with conform configuration. Running of conform manually is a hard way. Maybe we can add _run_hooks_from_dir "$PRE_CONFIGURE_HOOKS" in the boot script for `command`?

**Comment by bitwalker** [comment]
@umatomba I posted a comment in that thread, but yes, I plan to make functions available in the next release to run hooks for a desired phase (e.g. `run_hooks pre_start`).

**Comment by umatomba** [comment]
Thank you @bitwalker
